app/src/androidTest/java/ch/epfl/sdp/map/MapActivityTest.java

```diff
@@ -31,10 +40,71 @@
     @Before
     public void setUp() {
         mapFragment = (MapFragment) activityRule.getActivity().getFragment();
+        sentinel = new AtomicInteger(0);
     }
 
 
+    @Test(timeout = 30000)
+    public void testMapLoadCorrectly() throws Throwable {
+        while (mapFragment.getMap() == null){sleep(300);};
+
+        activityRule.runOnUiThread(() -> {
+            mapFragment.getMap().getStyle((s) -> sentinel.incrementAndGet());
+        }); // The sentinel value will only increase when the style has completely loaded
+
+        while (sentinel.get() == 0){sleep(300);}
+        sentinel.set(0);
+    }
+
 
+    @Test(timeout = 30000)
+    public void testHeatMapLoadCorrectly() throws Throwable {
+        testMapLoadCorrectly();
+
+        while (mapFragment.getHeatMapHandler() == null){sleep(300);};
+
+        activityRule.runOnUiThread(() -> {
+            mapFragment.getHeatMapHandler().onHeatMapDataLoaded(() -> sentinel.incrementAndGet());
+        }); // The sentinel value will only increase when the heatmap has completely loaded
+
+        while (sentinel.get() == 0){sleep(300);}
+        sentinel.set(0);
+    }
+
+    @Test(timeout = 50000)
+    public void testHeatMapToggleButton() throws Throwable {
+        testHeatMapLoadCorrectly();
+
+        mapFragment.onMapVisible(() -> sentinel.incrementAndGet());
+
+        while (sentinel.get() == 0){sleep(300);}
+        sentinel.set(0);
+
+        activityRule.runOnUiThread(() -> {
+            mapFragment.getMap().getStyle(style -> {
+                Layer layer = style.getLayer(HeatMapHandler.HEATMAP_LAYER_ID);
+                assertNotNull(layer);
+                assertEquals(VISIBLE, layer.getVisibility().getValue());
+            });
```

Comment:
You might want to consider extracting a method from this lambda used similarly in the next test :)

Reply:
Done in my last commit !